src/uct/tcp/tcp_ep.c

```diff
@@ -814,21 +818,20 @@ static inline ucs_status_t
 uct_tcp_ep_am_send(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
                    const uct_tcp_am_hdr_t *hdr)
 {
-    size_t sent_length;
-    int ret;
+    ssize_t ret;
 
     ep->tx.length       = sizeof(*hdr) + hdr->length;
     iface->outstanding += ep->tx.length;
 
-    ret = uct_tcp_ep_send(ep, &sent_length);
+    ret = uct_tcp_ep_send(ep);
     if (ucs_unlikely(UCS_STATUS_IS_ERR(ret))) {
```

Comment:
`ret <= 0`

Reply:
fixed to `ret < 0`. `ret == 0` is normal case

Reply:
only if send length = 0. how do you handle 'NO_PROGRESS'?

Reply:
if `NO_PROGRESS`, then `sent_length == 0`

Reply:
I meant input send length parameter, like `send(fd, NULL, /*length = */ 0, 0)` then it's OK but if `send(fd, buf, length /* != 0 */, 0)` returns 0 that's NO_PROGRESS. How do you differentiate these cases?

Reply:
BTW, read comment above as a pseudocode because `send(2)` will return -1 and errno = EAGAIN in second case

Reply:
> How do you differentiate these cases?

we don't need this
and also `ucs_socket_sendv_nb(length=0)` is impossible (we have assert for this)

Reply:
> BTW, read comment above as a pseudocode because `send(2)` will return -1 and errno = EAGAIN in second case

this will be returned by `ucs_socket_sendv_nb` as `sent_length=0` and `status=UCS_ERR_NO_PROGRESS`

Reply:
and this info is missed here, right?

Reply:
> and this info is missed here, right?

what info is missed here? do you mean to add some comment?

Reply:
> > How do you differentiate these cases?
> 
> we don't need this
> and also `ucs_socket_sendv_nb(length=0)` is impossible (we have assert for this)

I see only assertion for `iov_cnt`, number of bytes should be controlled by every implementation of `ucs_socket_iov_func_t iov_func`, so `ucs_socket_sendv_nb` can't guarantee this

Reply:
> > and this info is missed here, right?
> 
> what info is missed here? do you mean to add some comment?

You can't correctly handle `sent_length=0 and status=UCS_ERR_NO_PROGRESS` and `sent_length=0 and status=UCS_OK` cases

Reply:
sorry, this is for `ucs_socket_sendv_nb` only:
https://github.com/openucx/ucx/blob/d2f67c6e48e59be8b52504b41d9def135658e629/src/ucs/sys/sock.c#L243

TCP removes zero-length buffers from IOV. I agree we need to add `assert()` for `ucs_socket_sendv_nb()` to check that all IOV buffers are non-zero-length

Reply:
I don't think this is a right way, 0-byte messaging often is used for synchronization (not sure if it's possible for TCP)

Reply:
> You can't correctly handle `sent_length=0 and status=UCS_ERR_NO_PROGRESS` and `sent_length=0 and status=UCS_OK` cases

why should they be handled correctly from the caller's code?
Both of them are normal situations and don't require special handling from the caller's code.
They both mean that `0` bytes were sent and everything is good

Reply:
> ucp_worker_progress might return wrong value

we have `assert()` ensuring that data to be transmitted `> 0`:
https://github.com/openucx/ucx/blob/e07916faf31aff2251d08a8a8982d9f24026af3e/src/uct/tcp/tcp_ep.c#L499

So, in TCP code real send of `0` bytes is impossible. The only case `uct_tcp_ep_send` returns `0` is when `UCS_ERR_NO_PROGRESS` occurred.

Reply:
How does this assert ensure that TX data > 0? AFAIU it guarantee that sent data less or equal to buffer size

Reply:
> How does this assert ensure that TX data > 0?

```
sent_length = ep->tx.length - ep->tx.offset
``

`sent_length = 0` iff `ep->tx.length == ep->tx.offset`
but this is already handled in `assert(ep->tx.length > ep->tx.offset)`

Reply:
> AFAIU it guarantee that sent data less or equal to buffer size

yes, you're right - this check for this as well

Reply:
sent_length is in/out parameter, "out-value" can be LT "in-value"

Reply:
> sent_length is in/out parameter, "out-value" can be LT "in-value"

yes, right

Reply:
do you mean that `sent_length` "out-value" can be `0` and `status=UCS_OK`? and we should return `1` from `uct_iface_progress()`?

Reply:
I mean `ucs_assert(ep->tx.length > ep->tx.offset);` doesn't guarantee explicitly that "in TCP code real send of 0 bytes is impossible", even if it's really impossible due to other implicit statements on different levels of the stack.

Reply:
> I mean `ucs_assert(ep->tx.length > ep->tx.offset);` doesn't guarantee explicitly that "in TCP code real send of 0 bytes is impossible"

ok, let's assume that `ep->tx.length > ep->tx.offset`, e.g. `ep->tx.offset == 63` and `ep->tx.length == 64`
could you pls share how `sent_length` can become `0`?

Reply:
`ucs_assert(ep->tx.length > ep->tx.offset);` guarantees that we request to send non-zero number of bytes. In this case `0` can be returned in case of `NO_PROGRESS` only. Other errors would return something `<0`.

Reply:
I would like don't investigate this every time when see this code, e.g. during debugging of hardly reproducible bugs. This is possible according to the docs and API.
My original concern was about mixing return code approaches and possible missed cases. I was not going to block or require any changes for the PR but I can't approve this as well. I just pointed to possible issue. @brminich if you are OK with this I have no objections.

Reply:
> I would like don't investigate this every time when see this code, e.g. during debugging of hardly reproducible bugs. This is possible according to the docs and API.

What is possible? `sent_length == 0` when `ep->tx.length > ep->tx.offset`? How?

```
ucs_assert(ep->tx.length > ep->tx.offset);
sent_length = ep->tx.length - ep->tx.offset;
```

pls, provide an example when `ep->tx.length - ep->tx.offset` gives `0`